### ISSUE TITLE
function pointer type support in argspec

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#170]
It would be great if uftrace can provide function pointer type for argspec.
This can be useful when displaying some functions such as `pthread_create`.

**Comment by namhyung** [comment]
I wrote a code to do that and pushed to [review/arg-funcptr-v1](https://github.com/namhyung/uftrace/compare/review/arg-funcptr-v1).

It uses 'p' for function pointer like below:

```
$ uftrace -A pthread_create@arg3/p tests/t-thread-name
# DURATION    TID     FUNCTION
            [ 3131] | main() {
  84.233 us [ 3131] |   pthread_create(thread_first);
  28.096 us [ 3131] |   pthread_create(thread_second);
 309.918 us [ 3131] |   pthread_create(thread_third);
            [ 3131] |   pthread_create(thread_fourth) {
            [ 3136] | thread_second() {
            [ 3136] |   thread_first() {
   0.914 us [ 3136] |     foo();
   0.105 us [ 3136] |     bar();
   2.537 us [ 3136] |   } /* thread_first */
   3.700 us [ 3136] | } /* thread_second */
  62.011 us [ 3131] |   } /* pthread_create */
            [ 3131] |   pthread_join() {
            [ 3135] | thread_first() {
   0.937 us [ 3135] |   foo();
   0.112 us [ 3135] |   bar();
   3.420 us [ 3135] | } /* thread_first */
 263.040 us [ 3131] |   } /* pthread_join */
   1.037 us [ 3131] |   pthread_join();
            [ 3131] |   pthread_join() {
            [ 3137] | thread_third() {
            [ 3137] |   thread_first() {
   0.914 us [ 3137] |     foo();
   0.108 us [ 3137] |     bar();
   2.560 us [ 3137] |   } /* thread_first */
            [ 3138] | thread_fourth() {
   3.803 us [ 3137] | } /* thread_third */
            [ 3138] |   thread_first() {
   0.909 us [ 3138] |     foo();
   0.108 us [ 3138] |     bar();
   2.651 us [ 3138] |   } /* thread_first */
   3.868 us [ 3138] | } /* thread_fourth */
 125.444 us [ 3131] |   } /* pthread_join */
   1.240 us [ 3131] |   pthread_join();
 893.856 us [ 3131] | } /* main */
```

The thread_{second,third,fourth} don't call thread_first but it's optimized to jump to their.  With tail call handling, uftrace ends up showing them as function calls.

**Comment by honggyukim** [comment]
Thanks for doing this work! I think it looks great but what about changing the output from 'thread_first' to '(thread_first*)()' ?

**Comment by namhyung** [comment]
Wouldn't it hurt readability?

**Comment by honggyukim** [comment]
I'd like to show it as a function pointer shape. I thought about it again. What about `thread_first*`?would be better

**Comment by namhyung** [comment]
Hmm.. What about `&thread_first` then?

**Comment by honggyukim** [comment]
I think `&thread_first` is great. Thanks!

**Comment by namhyung** [comment]
Pushed [review/arg-funcptr-v2](https://github.com/namhyung/uftrace/compare/review/arg-funcptr-v2).

Changelog:
* prepend "&" sign to function name
* handle func-ptr arg in dump and script.
* add test case

**Comment by honggyukim** [comment]
Thanks! One more thing I want to discuss with you regarding the type symbol. Currently, we use a single type name as follows:
```
<format>      :=  "i" | "u" | "x" | "s" | "c" | "f" | "p" | "S"
```
But I'm thinking about adding more types such as std::vector, etc.  For example, "fp" for function pointer, "Sv" for std::vector, etc.
```
<format>      :=  "i" | "u" | "x" | "s" | "c" | "f" | "fp" | "S" | "Sv" | "Sl" | "Sd" ...
```
And we better think about enum type for future support.

**Comment by namhyung** [comment]
They are basically used for format specifier in printf(3) so I'd like to keep single character if possible.  If you could add support for those STL containers, then let's think about it again. But, to be honest, I'm a bit skeptical about the robustness.

**Comment by honggyukim** [comment]
Okay. Thanks for your comment. I also have to think about it again later regarding STL types. I'm okay with the current implementation with `p` then. Thanks!

**Comment by namhyung** [comment]
Merged: 2c595f1
